### PR TITLE
AEA-1119 sync connection and yoti connection implementation

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

sync connection and yoti connection

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/connections/sync_connection.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+            restricted_to_protocols=restricted_to_protocols,
+            excluded_protocols=excluded_protocols,
+            **kwargs,
+        )
+
+        self._tasks: Set[asyncio.Task] = set()
+
+    def _set_executor_pool(self, max_workers=None) -> None:
+        """Set executors pool."""
+        max_workers = max_workers or self.MAX_WORKER_THREADS
```
Comment (DavidMinarsch) [inline aea/connections/sync_connection.py:65]: should be configurable via `config` sub-field in yaml

File: aea/connections/sync_connection.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+from aea.configurations.base import ConnectionConfig
+from aea.configurations.data_types import PublicId
+from aea.connections.base import Connection, ConnectionStates
+from aea.crypto.wallet import CryptoStore
+from aea.identity.base import Identity
+from aea.mail.base import Envelope
+
+
+class BaseSyncConnection(Connection):
```
Comment (DavidMinarsch) [inline aea/connections/sync_connection.py:33]: Move the whole thing to `base.py`

File: aea/connections/sync_connection.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    @abstractmethod
+    def on_connect(self) -> None:
+        """Run on connect method called."""
+
+    @abstractmethod
+    def on_disconnect(self) -> None:
+        """Run on discconnect method called."""
+
+    @abstractmethod
+    def on_send(self, envelope: Envelope) -> None:
```
Comment (DavidMinarsch) [inline aea/connections/sync_connection.py:156]: Add another templated Sync Class in the existing scaffold.

File: aea/connections/base.py [hunk 4]
```diff
@@ -301,3 +301,143 @@ def is_connecting(self) -> bool:  # pragma: nocover
...
+        """Get an envelope from the connection."""
+        self._ensure_connected()
+        return await self._incoming_messages_queue.get()
+
+    def start_main(self) -> None:
+        """Start main function of the connection."""
+
+        def _():
+            task = self._loop.create_task(self._run_in_pool(self.main))
+            # cause task.set_name sfor python3.8+
```
Comment (DavidMinarsch) [inline aea/connections/base.py:423]:
```suggestion
            # cause task.set_name for python3.8+
```

File: aea/connections/base.py [hunk 5]
```diff
@@ -301,3 +301,143 @@ def is_connecting(self) -> bool:  # pragma: nocover
...
+            setattr(task, "task_name", "Connection main")  # noqa
+            task.add_done_callback(self._task_done_callback)
+            self._tasks.add(task)
+
+        self._loop.call_soon_threadsafe(_)
+
+    def main(self) -> None:
```
Comment (DavidMinarsch) [inline aea/connections/base.py:430]: Add `abstractmethod` decorator?

File: aea/connections/scaffold/connection.py [hunk 6]
```diff
@@ -16,16 +16,21 @@
...
 from aea.configurations.base import PublicId
-from aea.connections.base import Connection
+from aea.connections.base import BaseSyncConnection, Connection
 from aea.mail.base import Envelope
 
 
+"""
+Choose one of the possible implementations:
+
+Sync or Async connection and remove unused one.
```
Comment (DavidMinarsch) [inline aea/connections/scaffold/connection.py:30]:
```suggestion
Sync (inherited from BaseSyncConnection) or Async (inherited from Connection) connection and remove unused one.
```

File: aea/connections/scaffold/connection.py [hunk 7]
```diff
@@ -16,16 +16,21 @@
...
+"""
+
+
 class MyScaffoldConnection(Connection):
```
Comment (DavidMinarsch) [inline aea/connections/scaffold/connection.py:34]:
```suggestion
class MyScaffoldAsyncConnection(Connection):
```

File: aea/connections/scaffold/connection.py [hunk 8]
```diff
@@ -48,7 +53,7 @@ def __init__(self, **kwargs):
 
     async def connect(self) -> None:
         """
-        Set up the connection.
+        Tear down the connection.
```
Comment (DavidMinarsch) [inline aea/connections/scaffold/connection.py:56]: Revert

File: aea/connections/scaffold/connection.py [hunk 9]
```diff
@@ -78,3 +83,56 @@ async def receive(self, *args, **kwargs) -> Optional[Envelope]:
...
         """
         raise NotImplementedError  # pragma: no cover
+
+
+class MyScaffoldSyncConnection(BaseSyncConnection):
+    """Proxy to the functionality of the SDK or API."""
+
+    MAX_WORKER_THREADS = 5
+
+    connection_id = PublicId.from_str("fetchai/sync_scaffold:0.1.0")
```
Comment (DavidMinarsch) [inline aea/connections/scaffold/connection.py:93]: The public id would be the same, so `"fetchai/scaffold:0.1.0"`

File: aea/connections/scaffold/connection.py [hunk 10]
```diff
@@ -78,3 +83,56 @@ async def receive(self, *args, **kwargs) -> Optional[Envelope]:
...
+        :param identity: the identity object held by the agent.
+        :param crypto_store: the crypto store for encrypted communication.
+        :param restricted_to_protocols: the set of protocols ids of the only supported protocols for this connection.
+        :param excluded_protocols: the set of protocols ids that we want to exclude for this connection.
+        """
+        super().__init__(*args, **kwargs)
+        raise NotImplementedError  # pragma: no cover
+
+    def main(self):
+        """Run main."""
```
Comment (DavidMinarsch) [inline aea/connections/scaffold/connection.py:112]: Its not clear from the docstring what should be implemented here.

File: aea/connections/scaffold/connection.yaml [hunk 11]
```diff
@@ -8,7 +8,7 @@ license: Apache-2.0
 aea_version: '>=0.9.0, <0.10.0'
 fingerprint:
   __init__.py: QmZvYZ5ECcWwqiNGh8qNTg735wu51HqaLxTSifUxkQ4KGj
-  connection.py: QmUu1PRd5PJJk9Se7A6UH9TjTfDxhGKUHtRHgcsPz6xQb4
+  connection.py: QmeLAViUvvCKRpzVHADETJUr3G5XHDbUcK85t5NinajE5B
   readme.md: Qmdt71SaCCwAG1c24VktXDm4pxgUBiPMg4bWfUTiqorypf
 fingerprint_ignore_patterns: []
 connections: []
```
Comment (DavidMinarsch) [inline aea/connections/scaffold/connection.yaml:14]: update with `MyScaffoldAsyncConnection`

**Review by DavidMinarsch — COMMENTED** [review]
LGTM, I added some comments. There are also some test failures for Mac and some coverage gaps.

File: packages/fetchai/connections/yoti/__init__.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""Scaffold of a connection."""
```
[the remark this answers is not in the document]
Comment (solarw) [reply packages/fetchai/connections/yoti/__init__.py:20]: done

File: aea/connections/base.py (hunk 5 above)
Comment (solarw) [reply aea/connections/base.py:431]: no, sometimes main is not required, see no reason to force develope to define empty main if not needed

File: aea/connections/scaffold/connection.yaml (hunk 11 above)
Comment (solarw) [reply aea/connections/scaffold/connection.yaml:14]: done

File: aea/connections/scaffold/connection.py (hunk 10 above)
Comment (solarw) [reply aea/connections/scaffold/connection.py]: updated

File: aea/connections/scaffold/connection.py (hunk 8 above)
Comment (solarw) [reply aea/connections/scaffold/connection.py]: done

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**